flowmachine/flowmachine/features/subscriber/distance_series.py

```diff
@@ -0,0 +1,140 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# -*- coding: utf-8 -*-
+"""
+Per subscriber time series of distances from some reference location.
+"""
+from typing import List, Optional
+
+from flowmachine.features.spatial import DistanceMatrix
+from .metaclasses import SubscriberFeature
+from ..utilities.subscriber_locations import SubscriberLocations, BaseLocation
+
+
+valid_stats = {"sum", "avg", "max", "min", "median", "stddev", "variance"}
+valid_time_buckets = ["second", "minute", "hour", "day", "month", "year", "century"]
+
+
+class DistanceSeries(SubscriberFeature):
+    """
+    Per subscriber time series of distance in meters from some reference location.
+
+    Parameters
+    ----------
+    subscriber_locations : SubscriberLocations
+        A subscriber locations query with a lon-lat spatial unit to build the distance series against.
+    reference_location : BaseLocation or None, default None
+        The set of home locations from which to calculate distance at each sighting.
+        If not given then distances will be from (0, 0).
+    statistic : str
+        the statistic to calculate one of 'sum', 'avg', 'max', 'min', 
+        'median', 'stddev' or 'variance'
+    time_bucket : {"second", "minute", "hour", "day", "month", "year", "century"}, default "day"
+        Time bucket to calculate the statistic over.
+
+    Examples
+    --------
+    >>> d = DistanceSeries(subscriber_locations=SubscriberLocations("2016-01-01", "2016-01-07", spatial_unit=make_spatial_unit("lon-lat")))
+    >>> d.head()
+             subscriber    datetime         value
+    0  038OVABN11Ak4W5P  2016-01-01  9.384215e+06
+    1  038OVABN11Ak4W5P  2016-01-02  9.233302e+06
+    2  038OVABN11Ak4W5P  2016-01-03  9.376996e+06
+    3  038OVABN11Ak4W5P  2016-01-04  9.401404e+06
+    4  038OVABN11Ak4W5P  2016-01-05  9.357210e+06
+    """
+
+    def __init__(
+        self,
+        *,
+        subscriber_locations: SubscriberLocations,
+        reference_location: Optional[BaseLocation] = None,
+        statistic: str = "avg",
+        time_bucket: str = "day",
+    ):
+        subscriber_locations.spatial_unit.verify_criterion("has_geography")
+        subscriber_locations.spatial_unit.verify_criterion("has_lon_lat_columns")
+        self.spatial_unit = subscriber_locations.spatial_unit
+        if time_bucket.lower() in valid_time_buckets:
+            self.aggregate_by = time_bucket.lower()
+        else:
+            raise ValueError(
+                f"'{time_bucket}' is not a valid value for time_bucket. Use one of {valid_time_buckets}"
```

Comment:
I realise that we often perform ad-hoc validations whenever a parameter is passed (not just in this here, also elsewhere in the code base). In addition, we often use strings as a replacement for more meaningful types but then also add extra behaviour (e.g. casting the time bucket to a date depending on the granularity). It may be worth creating explicit classes which can absorb this behaviour (e.g `TimeBucket` and `AggregationStatistic` for the strings used in this file). Definitely not needed for this PR, just as a thought for the future.

Reply:
An advantage of having a dedicated class for this would be that we don't need to repeat the valid values in the docstring here (or anywhere else where it is used) but can just refer to the docstring of `TimeBucket`/`AggregationStatistic`.

Reply:
(And it would allow us to get rid of various tests for invalid time buckets / statistics, which are currently probably repeated in multiple places, because invalid values would only be tested once for those classes.)

Reply:
Again, definitely not for this PR. I'll probably add separate Github issues for this so that we can discuss/track it there.

Reply:
Agree - one nice thing about a class for time buckets would be that one could do ordered comparisons between them as well.